File: src/javascript/algorithms/dynamicProgramming/wordBreak.js
//TC: O(n*2 *m) SC:O(n)

function wordBreak(str, wordDict) {
    let dp = new Array(str.length+1).fill(false);
    dp[str.length] = true;
    for(let i=str.length-1; i>=0; i--) {
        for(let word of wordDict) {
            if((i+word.length <= str.length) && wordDict.includes(str.substr(i, word.length))) {
                dp[i] = dp[i+ word.length];
            }
            if(dp[i]) {
                break;
            }
        }
    }
    return dp[0];
}

let str = "applepenapple";
let wordDict = ["pen", "apple"];
let str1 = "catsandog", wordDict1 = ["cats","dog","sand","and","cat"]
console.log(wordBreak(str, wordDict));
console.log(wordBreak(str1, wordDict1));

File: src/java1/algorithms/dynamicProgramming/wordBreak/WordBreak.java
package java1.algorithms.dynamicProgramming.wordBreak;
import java.util.*;

//1D Bottom-up DP with reverse iteration => TC: O(n * m * k) SC:O(n) n=number of chars in string, m = number of words in dict, k=Avg word length
public class WordBreak {
    private static boolean wordBreak(String str, List<String> wordDict) {
            boolean[] dp = new boolean[str.length()+1];
            dp[str.length()] = true;
            for(int i = str.length()-1; i>=0; i--) {
                for(String word: wordDict) {
                    if((i+ word.length() <= str.length()) && str.startsWith(word, i)) {
                        if(dp[i+ word.length()]) {
                            dp[i] = dp[i+ word.length()];
                            break;
                        }
                    }
                }
            }
            return dp[0];
    }

    public static void main(String[] args) {
        String str1 = "applepenapple";
        List<String> wordDict1 = Arrays.asList("pen", "apple");
        String str2 = "catsandog";
        List<String> wordDict2 = Arrays.asList("cats", "dog", "and", "cat");
        System.out.println(wordBreak(str1, wordDict1));
        System.out.println(wordBreak(str2, wordDict2));
    }
}
